### PR TITLE
*: expose backend and refactor backend files

### PR DESCRIPTION
**Pull request by fanminshi**
expose backend interface and backend objects. 
refactor any backend related files into backup/backend dir.

### CODE REVIEW

File: pkg/backup/backend/abs.go [hunk 1]
```diff
@@ -12,27 +12,29 @@
...
 	"github.com/sirupsen/logrus"
 )
 
 // ensure absBackend satisfies backend interface.
-var _ backend = &absBackend{}
+var _ Backend = &AbsBackend{}
 
-type absBackend struct {
+// AbsBackend is the Azure Blob Storage backend.
+type AbsBackend struct {
```
Comment (hongchaodeng) [inline pkg/backup/backend/abs.go]: this should be private

File: pkg/backup/backend/backend.go [hunk 2]
```diff
@@ -12,27 +12,28 @@
...
 
 import "io"
 
-type backend interface {
+// Backend defines required backend operations
+type Backend interface {
 	// save saves the backup from the given reader with given etcd version and revision.
 	// It returns the size of the snapshot saved.
-	save(etcdVersion string, rev int64, r io.Reader) (size int64, err error)
+	Save(etcdVersion string, rev int64, r io.Reader) (size int64, err error)
```
Comment (hongchaodeng) [inline pkg/backup/backend/backend.go]: Update the doc string as well?

File: pkg/backup/backend/abs.go (hunk 1 above)
Comment (fanminshi) [reply pkg/backup/backend/abs.go]: it must be exposed because backup.go in backup pkg uses it.

File: pkg/backup/constants/constants.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package constants
+
+const (
+	BackupTmpDir         = "tmp"
```
Comment (hongchaodeng) [inline pkg/backup/constants/constants.go]: No need to create new pkg. Just put them into backup util

File: pkg/backup/backend/backend.go (hunk 2 above)
Comment (fanminshi) [reply pkg/backup/backend/backend.go]: good catch.

File: pkg/backup/backend/abs_backend_test.go [hunk 4]
```diff
@@ -25,6 +25,7 @@ import (
 
 	"github.com/Azure/azure-sdk-for-go/storage"
```
Comment (hasbro17) [inline pkg/backup/backend/abs_backend_test.go]: Separate imports.

File: pkg/backup/backend/s3_backend_test.go [hunk 5]
```diff
@@ -27,6 +27,7 @@ import (
 	"github.com/aws/aws-sdk-go/aws/credentials"
 	"github.com/aws/aws-sdk-go/aws/session"
```
Comment (hasbro17) [inline pkg/backup/backend/s3_backend_test.go]: Separate imports.

File: pkg/backup/backup.go [hunk 6]
```diff
@@ -27,10 +27,12 @@ import (
 	"github.com/coreos/etcd-operator/pkg/backup/backupapi"
 	"github.com/coreos/etcd-operator/pkg/backup/env"
 	"github.com/coreos/etcd-operator/pkg/backup/s3"
+	"github.com/coreos/etcd-operator/pkg/backup/util"
 	"github.com/coreos/etcd-operator/pkg/util/constants"
 	"github.com/coreos/etcd-operator/pkg/util/etcdutil"
 	"github.com/coreos/etcd-operator/pkg/util/k8sutil"
 
+	"github.com/coreos/etcd-operator/pkg/backup/backend"
```
Comment (hasbro17) [inline pkg/backup/backup.go]: Separate imports.

**Comment by hasbro17** [comment]
LGTM after CI passes.

**Pull request merged**